security/tpm/test/_tpm_test.js

```diff
@@ -60,26 +87,56 @@ describe('tpm', function () {
         done();
       });
     });
+
+    it('creates an endorsement key if one does not exist', function(done) {
+      /*Tests_SRS_NODE_TPM_SECURITY_CLIENT_06_017: [If the endorsement key does NOT exist, a new key will be created.] */
+      var tpm = new tss.Tpm();
+      var connectStub = sinon.stub(tpm, 'connect');
+      connectStub.callsArgWith(0);
+      var client = new TpmSecurityClient(undefined, tpm);
+      var readPublicStub = sinon.stub(tpm, 'ReadPublic');
+      var lastResponseStub = sinon.stub(tpm, 'getLastResponseCode');
+      lastResponseStub.returns(tss.TPM_RC.SUCCESS);
+      lastResponseStub.onCall(0).returns(tss.TPM_RC.BAD_TAG);
+      var createPrimaryStub = sinon.stub(tpm, 'CreatePrimary');
+      createPrimaryStub.callsArgWith(5, fakeCreatedEkKey);
+      var evictControlStub = sinon.stub(tpm, 'EvictControl').callsArg(3);
+      var flushContextStub = sinon.stub(tpm, 'FlushContext').callsArg(1);
+      readPublicStub.withArgs(TpmSecurityClient._ekPersistentHandle).callsArgWith(1, fakeEkKey);
+      readPublicStub.withArgs(TpmSecurityClient._srkPersistentHandle).callsArgWith(1, fakeSrkKey);
+      readPublicStub.withArgs(TpmSecurityClient._idKeyPersistentHandle).callsArgWith(1, fakeIdKey);
+      client.getEndorsementKey((err, localEk) => {
+        assert.deepEqual(localEk, fakeCreatedEkKey.outPublic.asTpm2B(), 'Invalid endorsement key returned.');
+        done();
+      });
+    });
   });
 
   describe('getStorageRootKey', function() {
     it('returns the storage root key', function(done) {
       /*Tests_SRS_NODE_TPM_SECURITY_CLIENT_06_008: [The `getStorageRootKey` function shall query the TPM hardware and return the `storageRootKey` in the callback.] */
-      var client = new TpmSecurityClient(undefined, fakeSimpleTpmClient );
-      var createPersistStub = sinon.stub(client,'_createPersistentPrimary');
-      var readPersistStub = sinon.stub(client,'_readPersistentPrimary');
-      createPersistStub.withArgs('EK').callsArgWith(4, null, TpmSecurityClient._ekTemplate);
-      createPersistStub.withArgs('SRK').callsArgWith(4, null, TpmSecurityClient._srkTemplate);
-      readPersistStub.withArgs('IDENTITY').callsArgWith(2, null, TpmSecurityClient._srkTemplate)
+      var tpm = new tss.Tpm();
+      var connectStub = sinon.stub(tpm, 'connect');
+      connectStub.callsArgWith(0);
+      var client = new TpmSecurityClient(undefined, tpm);
+      var readPublicStub = sinon.stub(tpm, 'ReadPublic');
+      var lastResponseStub = sinon.stub(tpm, 'getLastResponseCode');
+      lastResponseStub.returns(tss.TPM_RC.SUCCESS);
+      readPublicStub.withArgs(TpmSecurityClient._ekPersistentHandle).callsArgWith(1, fakeEkKey);
+      readPublicStub.withArgs(TpmSecurityClient._srkPersistentHandle).callsArgWith(1, fakeSrkKey);
+      readPublicStub.withArgs(TpmSecurityClient._idKeyPersistentHandle).callsArgWith(1, fakeIdKey);
```

Comment:
I'm not sure if the same stubs are used over and over everywhere but if it's the case maybe refactoring them in a single function (createStubbedClient or something along those lines) would help?

Reply:
Not always quite the same.

---
In reply to: [167385775](https://github.com/Azure/azure-iot-sdk-node/pull/225#discussion_r167385775) [](ancestors = 167385775)